modules/gis/src/lib/geojson-to-flatGeojson.ts

```diff
@@ -0,0 +1,122 @@
+import {getPolygonSignedArea} from '@math.gl/polygon';
+
+import {Feature, Position, FlatFeature} from '@loaders.gl/schema';
+
+export type GeojsonToFlatGeojsonOptions = {
+  coordLength: number;
+  fixRingWinding: boolean;
+};
+
+export function geojsonToFlatGeojson(
+  features: Feature[],
+  options: GeojsonToFlatGeojsonOptions = {coordLength: 2, fixRingWinding: true}
+): FlatFeature[] {
+  return features.map((feature) => flattenFeature(feature, options));
+}
+
+function flattenPoint(
+  coordinates: Position,
+  data: number[],
+  lines: number[],
+  options: GeojsonToFlatGeojsonOptions
+) {
+  lines.push(data.length);
+  data.push(...coordinates);
+
+  // Pad up to coordLength
+  for (let i = coordinates.length; i < options.coordLength; i++) {
+    data.push(0);
+  }
+}
+
+function flattenLineString(
+  coordinates: Position[],
+  data: number[],
+  lines: number[],
+  options: GeojsonToFlatGeojsonOptions
+) {
+  lines.push(data.length);
+  for (const c of coordinates) {
+    data.push(...c);
+
+    // Pad up to coordLength
+    for (let i = c.length; i < options.coordLength; i++) {
+      data.push(0);
+    }
+  }
+}
+
+function flattenPolygon(
+  coordinates: Position[][],
+  data: number[],
+  lines: number[][],
+  areas: number[][],
+  options: GeojsonToFlatGeojsonOptions
+) {
+  let count = 0;
+  const ringAreas: number[] = [];
+  const polygons: number[] = [];
+  for (const lineString of coordinates) {
+    const lineString2d = lineString.map((p) => p.slice(0, 2));
+    let area = getPolygonSignedArea(lineString2d.flat());
+    const ccw = area < 0;
+
+    // Exterior ring must be CCW and interior rings CW
+    if (options.fixRingWinding && ((count === 0 && !ccw) || (count > 0 && ccw))) {
+      lineString.reverse();
+      area = -area;
+    }
+    ringAreas.push(area);
+    flattenLineString(lineString, data, polygons, options);
+    count++;
+  }
+
+  if (count > 0) {
+    areas.push(ringAreas);
+    lines.push(polygons);
+  }
+}
+
+function flattenFeature(feature: Feature, options: GeojsonToFlatGeojsonOptions): FlatFeature {
+  const {geometry} = feature;
+  if (geometry.type === 'GeometryCollection') {
+    throw new Error('GeometryCollection type not supported');
+  }
+  const data = [];
+  const lines = [];
+  let areas;
+  let type;
+
+  switch (geometry.type) {
+    case 'Point':
+      type = 'Point';
+      flattenPoint(geometry.coordinates, data, lines, options);
```

Comment:
Just wondering: why do you have these set up to mutate `data` and `lines` in the upper scope instead of just returning those from the `flatten*` function? For performance?

Reply:
To avoid creating many arrays and putting pressure on garbage collection is the practical reason. But I also find it more readable. ` geometry.coordinates.map((c) => flattenPolygon(c, data, lines, areas, options));` seems tidier than `geometry.coordinates.map((c) => [data, lines, areas] = flattenPolygon(c, data, lines, areas, options));`